File: writers/markdown_writer.go
package writers

import (
	"fmt"
	"strings"

	"github.com/naoty/table/table"
	"github.com/naoty/table/utils"
)

// MarkdownWriter is a Writer to draw tables in markdown format.
type MarkdownWriter struct {
}

// Write makes a string for a given table in markdown format.
func (writer MarkdownWriter) Write(table *table.Table) string {
	columnWidths := []int{}
	for _, column := range table.Columns {
		columnWidths = append(columnWidths, column.Width())
	}

	buf := []string{}

	// Header
	if table.Header != nil {
		buf = append(buf, writer.WriteRow(table.Header, columnWidths))
		buf = append(buf, writer.WriteBorder(table, columnWidths))
	}

	// Each rows
	for _, row := range table.Rows {
		buf = append(buf, writer.WriteRow(row, columnWidths))
	}

	return fmt.Sprintf("%v\n", strings.Join(buf, "\n"))
}

// WriteRow makes string represents a row with given column widths.
func (writer MarkdownWriter) WriteRow(row table.Row, widths []int) string {
	buf := []string{"|"}

	for i, item := range row {
		buf = append(buf, " ")
		buf = append(buf, item)

		paddingWidth := widths[i] - utils.LenInHalfSize(item)
		buf = append(buf, strings.Repeat(" ", paddingWidth))
		buf = append(buf, " |")
	}

	return strings.Join(buf, "")
}

// WriteBorder makes string represents a border with given column widths.
func (writer MarkdownWriter) WriteBorder(table *table.Table, widths []int) string {
	buf := []string{}

	for _, width := range widths {
		buf = append(buf, strings.Repeat("-", width))
	}

	return "| " + strings.Join(buf, " | ") + " |"
}

File: src/writer/markdown_writer.rs
use std::io;
use unicode_width::UnicodeWidthStr;

pub struct MarkdownWriter<T: io::Write> {
  writer: T,
}

impl<T: io::Write> MarkdownWriter<T> {
  pub fn new(writer: T) -> Self {
    MarkdownWriter { writer }
  }
}

impl<T: io::Write> crate::Write for MarkdownWriter<T> {
  fn write(&mut self, table: crate::Table) -> crate::Result<()> {
    let column_widths = table.column_widths();

    let mut border = String::new();
    for column_width in column_widths.iter() {
      border += &format!("| {} ", "-".repeat(column_width.clone()));
    }
    border += "|\n";

    for (i, row) in table.rows.iter().enumerate() {
      for (j, value) in row.iter().enumerate() {
        let spaces = " ".repeat(column_widths[j] - UnicodeWidthStr::width(value as &str));
        let cell = format!("| {}{} ", value, spaces);
        self.writer.write(cell.as_bytes())?;
      }

      self.writer.write(b"|\n")?;

      if i == 0 {
        self.writer.write(border.as_bytes())?;
      }
    }

    Ok(())
  }

  fn flush(&mut self) -> crate::Result<()> {
    self.writer.flush()?;
    Ok(())
  }
}
